internal/pkg/service/common/etcdop/mutex.go

```diff
@@ -0,0 +1,347 @@
+			return err
+		}
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	l.locked = mtx
+	return nil
+}
+
+// TryLock locks the mutex if not already locked by another session.
+// If lock is held by another session, return immediately after attempting necessary cleanup
+// Internally, the lock is backed by a local lock and an etcd mutex,
+// so the protected operation runs only once within the entire cluster.
+//
+// You cannot use etcd as a naive locking system:
+// carefully couple a fencing token, read more: https://github.com/etcd-io/etcd/issues/11457
+func (l *Mutex) TryLock(ctx context.Context) error {
+	mtx, err := l.store.lock(ctx, l.name, func(mtx *activeMutex) error {
+		// Try to acquire the local lock
+		if err := mtx.localTryLock(ctx); err != nil {
+			return err
+		}
+
+		// Try to acquire the DB lock, only at the first time
+		if err := mtx.dbTryLock(ctx); err != nil {
+			_ = mtx.localUnlock() // revert the local lock on error
+			return err
+		}
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	l.locked = mtx
+	return nil
+}
+
+func (l *Mutex) Unlock(ctx context.Context) error {
+	if l.locked == nil {
+		return NotLockedError{}
+	}
+
+	err := l.locked.unlock(ctx)
+	l.locked = nil
+	return err
+}
+
+func (s *mutexStore) NewMutex(name string) *Mutex {
+	return &Mutex{store: s, name: name}
+}
+
+// get returns a mutex with the name.
+func (s *mutexStore) get(name string) *activeMutex {
+	s.allLock.Lock()
+	defer s.allLock.Unlock()
+
+	mtx, ok := s.all[name]
+	if ok {
+		// Increment usage, see clear method
+		mtx.usage++
+		return mtx
+	}
+
+	// Create mutex
+	mtx = &activeMutex{store: s, name: name, usage: 1, localMutex: make(chan struct{}, 1)}
+	s.all[name] = mtx
+	return mtx
+}
+
+// getReady returns a mutex with the name, it the mutex is in teardown state, the operation is retried.
+func (s *mutexStore) getReady(name string) *activeMutex {
+	for {
+		// Get or create mutex
+		mtx := s.get(name)
+
+		// Check if the mutex is ready
+		if mtx.teardown == nil {
+			return mtx
+		}
+
+		// Wait until the mutex in unlocked in DB, so it can be locked again, see clear method
+		<-mtx.teardown
+	}
+}
+
+// clear the mutex, if it is no more used.
+func (s *mutexStore) clear(ctx context.Context, mtx *activeMutex) (err error) {
+	s.allLock.Lock()
+	clearMtx := false
+	unlockDB := false
+	if mtx.usage--; mtx.usage == 0 {
+		clearMtx = true
+		unlockDB = mtx.dbMutex != nil
+		mtx.teardown = make(chan struct{})
+	}
+	s.allLock.Unlock()
+
+	// Unlock DB if the lock is no more used
+	if unlockDB {
+		if err = mtx.dbMutex.Unlock(ctx); err != nil {
+			// DB connection is not working, state of the lock is unknown, close session
+			_ = mtx.dbSession.Close()
+		}
+	}
+
+	// Clear memory
+	if clearMtx {
+		s.allLock.Lock()
+		mtx.dbSession = nil
+		mtx.dbMutex = nil
+		delete(s.all, mtx.name) // prevent memory leak, remove unused mutex from the map
+		close(mtx.teardown)
+		s.allLock.Unlock()
+	}
+
+	return err
+}
```

Comment:
This is the hard part, we should prevent memory leaks.

The lock is for example a `FileKey.String()`, so, it is used several times and then should be removed from the map.

But during removal/clear in the database, a request for the same lock may come (edge-case).

So the new request must wait for the DB unlock, a then a new lock can be acquired.